### ISSUE TITLE
Some conditions cause type inference to fail after if statement

### ISSUE THREAD

**Issue by SamMousa** [vimeo/psalm#8426]
I have code like this:

```php
function(array|string $b) {
    if (is_array($b) || $someCondition) {
        return;

    }
    // Here phpstan knows that `$b` is not an array.

```

However depending on the type of condition on the right of the `||` operator, sometimes it doesn't know. This is weird since adding more alternatives makes the guard weaker not stronger...

https://psalm.dev/r/fcab2f139f

**Comment by orklah** [comment]
There comes a point where Psalm can't process what you throw at it anymore. It could certainly be improved, possibly by still processing part of the condition when it's possible.

**Comment by SamMousa** [comment]
I understand that; but I don't get why the reasoning fails. It feels like a bug to me. Check the 2 examples in the code, only difference is a `!`

**Comment by SamMousa** [comment]
https://psalm.dev/r/4bce63f076

Replacing `!isset(` with `false === isset(` fixes it :|

**Comment by orklah** [comment]
Yeah, this would need to be looked at. I still think it's Psalm dropping a complex expression, but if the false === is working, it may be possible to make it work
